MdeModulePkg/Core/Dxe/Misc/MemoryProtectionSupport.c

```diff
@@ -2941,21 +2954,80 @@ MapLegacyBiosMemoryRWX (
   VOID
   )
 {
-  EFI_STATUS  Status = EFI_SUCCESS;
+  EFI_STATUS                       Status = EFI_SUCCESS;
+  EFI_GCD_MEMORY_SPACE_DESCRIPTOR  Desc;
+  EFI_PHYSICAL_ADDRESS             Start = 0x0;
+  UINT64                           Length;
 
-  // https://wiki.osdev.org/Memory_Map_(x86)
-  //
-  // Map the legacy BIOS write-back memory as RWX.
-  if (gCpu != NULL) {
+  if (gCpu == NULL) {
+    DEBUG ((
+      DEBUG_ERROR,
+      "%a cannot remap Legacy BIOS memory as RWX because gCpu is NULL\n",
+      __func__
+      ));
+    return;
+  }
+
+  // Ensure that this memory is marked as system memory. If it is not system memory, do not change
+  // the memory attributes as we do not want to map something that shouldn't be mapped or map something
+  // incorrectly
+  while (Start < LEGACY_BIOS_WB_LENGTH) {
+    Status = CoreGetMemorySpaceDescriptor (Start, &Desc);
+    if (EFI_ERROR (Status)) {
+      DEBUG ((
+        DEBUG_ERROR,
+        "%a - Failed to get memory space descriptor for address 0x%llx! Status = %r\n",
+        __func__,
+        Start,
+        Status
+        ));
+      return;
+    }
+
+    if (Desc.BaseAddress + Desc.Length > LEGACY_BIOS_WB_LENGTH) {
+      Length = LEGACY_BIOS_WB_LENGTH - Desc.BaseAddress;

```

Comment:
Can you explain how this Length variable differs and why it's not used on line 3003?

Reply:
`Length` here is the range of memory that we are marking as RWX. `Start` is being used to look through the different memory descriptors (which may be one for multiple and may extend past the end of the range we need to mark RWX). So in 3003 and 3029 we are navigating the descriptors, not the length we are changing attributes for.